distributed/service.py

```diff
@@ -0,0 +1,426 @@
+array rechunking, etc.) are inefficient when represented as task graphs, but can be
+implemented simply and performantly as bespoke systems that handle their own low-level
+data transfer and computation manually.
+
+The `Service` interface offers a structured way to "hand off" data from the task-graph
+paradigm to these out-of-band operations, and to "hand back" the results so downstream
+tasks can use them, while maintaining resilience of the overall graph.
+
+In a graph like::
+
+    {
+        ("input", 1): 1,
+        ("input", 2): 2,
+        "foo-service": (FooService, [("input", 1), ("input", 2)]),  # annotated as a service
+        ("output", 1): (service_result, "foo-service"),             # never actually runs
+        ("output", 2): (service_result, "foo-service"),             # never actually runs
+        ("output", 3): (service_result, "foo-service"),             # never actually runs
+        ("downstream", 1): (use, ("output", 1)),
+        ("downstream", 2): (use, ("output", 2)),
+        ("downstream", 3): (use, ("output", 3)),
+    }
+
+Or visually::
+
+    d1   d2   d3
+    |    |    |
+    o1   o2   o3
+      \  |   /
+    foo-service
+        / \
+      i1  i2
+
+If the single ``foo-service`` task is annotated as a service, it'll actually be computed
+by launching ``FooService`` instances on every worker, passing the inputs into them,
+letting them do whatever out-of-band operations they wish (including communicating with
+each other or external systems), and waiting for them to produce the output keys. Once
+each output key is produced (the `service_result` functions never actually run), any
+tasks downstream of it can run.
+
+---------------------------------------------------------------------------------------
+
+Dask handles startup/shutdown, resilience, peer discovery, leader election, RPC
+interfaces, and composition for Services:
+
+Startup/shutdown
+----------------
+
+* Once the *first* input key to a service task is ready (state ``memory``), service
+  instances are started on all workers in the cluster.
+* Once the last output key is produced by a service, all instances are told to stop.
+* Any instance can error the service task; all other instances will be told to stop.
+* Any instance can restart the service task; all other instances will be told to stop,
+  then re-launched from scratch.
+* If the service task is cancelled (because keys downstream from it are released, etc.),
+  all instances are told to stop.
+
+Resilience
+----------
+
+* If a worker that's running a service leaves, the entire service task is restarted on
+  all workers. TODO change once `peer_joined`/`peer_left` are added.
+* If keys downstream of a service need to be recomputed, the service will be rerun, just
+  like regular tasks are.
+* The service is informed at runtime which inputs to expect and which outputs to
+  produce, so partial recomputation (some outputs already in memory, some not) can be
+  handled correctly.
```

Comment:
Maybe related to my other comment on this line.

Would this solve our "culling" problem? If so, how? How would the scheduler detect that the graph is culled and how would the information be propagated to the services?